nav2_amcl/src/amcl_node.cpp

```diff
@@ -20,440 +20,219 @@
+  tf_listener_.reset();
+  tf_buffer_.reset();
 
-void
-AmclNode::handleMapMessage(const nav_msgs::msg::OccupancyGrid & msg)
-{
-  std::lock_guard<std::recursive_mutex> cfl(configuration_mutex_);
+  // PubSub
+  pose_pub_.reset();
+  particlecloud_pub_.reset();
 
-  RCLCPP_INFO(get_logger(), "Received a %d X %d map @ %.3f m/pix\n",
-    msg.info.width,
-    msg.info.height,
-    msg.info.resolution);
-  if (msg.header.frame_id != global_frame_id_) {
-    RCLCPP_WARN(
-      get_logger(), "Frame_id of map received:'%s' doesn't match global_frame_id:'%s'. This could"
-      " cause issues with reading published topics",
-      msg.header.frame_id.c_str(),
-      global_frame_id_.c_str());
-  }
+  // Odometry
+  motion_model_.reset();
 
-  freeMapDependentMemory();
-  // Clear queued laser objects because they hold pointers to the existing
-  // map, #5202.
+  // Particle Filter
+  pf_free(pf_);
+  pf_ = nullptr;
+
+  // Laser Scan
   lasers_.clear();
   lasers_update_.clear();
   frame_to_laser_.clear();
 
-  map_ = convertMap(msg);
-
-#if NEW_UNIFORM_SAMPLING
-  // Index of free space
-  free_space_indices.resize(0);
-  for (int i = 0; i < map_->size_x; i++) {
-    for (int j = 0; j < map_->size_y; j++) {
-      if (map_->cells[MAP_INDEX(map_, i, j)].occ_state == -1) {
-        free_space_indices.push_back(std::make_pair(i, j));
-      }
-    }
-  }
-#endif
-  // Create the particle filter
-  pf_ = pf_alloc(min_particles_, max_particles_,
-      alpha_slow_, alpha_fast_,
-      (pf_init_model_fn_t)AmclNode::uniformPoseGenerator,
-      reinterpret_cast<void *>(map_));  // (void *)map_);
-  pf_->pop_err = pf_err_;
-  pf_->pop_z = pf_z_;
-
-  // Initialize the filter
-  updatePoseFromServer();
-  pf_vector_t pf_init_pose_mean = pf_vector_zero();
-  pf_init_pose_mean.v[0] = init_pose_[0];
-  pf_init_pose_mean.v[1] = init_pose_[1];
-  pf_init_pose_mean.v[2] = init_pose_[2];
-  pf_matrix_t pf_init_pose_cov = pf_matrix_zero();
-  pf_init_pose_cov.m[0][0] = init_cov_[0];
-  pf_init_pose_cov.m[1][1] = init_cov_[1];
-  pf_init_pose_cov.m[2][2] = init_cov_[2];
-  pf_init(pf_, pf_init_pose_mean, pf_init_pose_cov);
-  pf_init_ = false;
-
-  // Instantiate the sensor objects
-  delete motionModel_;
-  createMotionModel();
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
 
-  // Laser
-  delete laser_;
-  createLaserObject();
+nav2_lifecycle::CallbackReturn
+AmclNode::on_error(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_ERROR(get_logger(), "Handling error state");
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
 
-  // In case the initial pose message arrived before the first map,
-  // try to apply the initial pose now that the map has arrived.
-  applyInitialPose();
+nav2_lifecycle::CallbackReturn
+AmclNode::on_shutdown(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
 }
 
 void
-AmclNode::freeMapDependentMemory()
+AmclNode::checkLaserReceived()
```

Comment:
In the original code, there was a `laser_check_interval_` timer that was checking to see if a laser scan is received within some time.  Perhaps we can do the same thing here. If we don't receive a laser data within some time, we can bring the node into inactive state and provide warning.

Reply:
@crdelsey Yes, I left this code, intending that we discuss this issue. Lifecycle nodes can check their prerequisites during on_configure (or on_activate). How do we want them to behave on an ongoing basis (during the Active state)? For example, we could add a timer to our own LifecycleNode base class and a checkPrerequisites function that the derived class can override. However, this might be an issue for real-time/deterministic behavior. Anyways, I'd like to address this issue in a general way and apply it to all of our modules. For now, I can remove this code if you prefer.

Reply:
I'd rather we added back in the wall timer, rather than drop the functionality.